### PR TITLE
`Programming exercises`: Cache dependencies in the Java blackbox template

### PR DESCRIPTION
**Pull request by BaumiCoder**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Client
No Changes at the client

#### Changes affecting Programming Exercises
- [x] ~~**High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).~~ Only applicable to the GitLab/Jenkins template. The Aeolus/LocalCI one does not support setting Docker run flags specific to the exercise.
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.

### Motivation and Context

The actual caching approach with Docker volumes has one mature disadvange.
It needs to have a testing run for the solution before a testing run for a student submission
on each host, which is running the containers.
That is a problem if you need more hosts to run the container because of scaling issues.

The new approach uses a dedicated cache server (Maven proxy) that does not have this limitation.

### Description

1. Switch the blackbox Dejagnu template to cache Maven dependencies with Sonatype Nexus.
2. Document how to use the Sonatype Nexus Caching for Maven and Gradle.
3. Document how to restrict / disable the network access for student submission corresponding to the used caching approach.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- A course where the student is enrolled / can enrol

1. Log in to Artemis with the Instructor account
    1. Navigate to a course
    2. Create a new Programming Exercise there
         1.  Choose language `Java` 
         2.  Choose type `DejaGnu`
    3. Check if the tests for the solution repository are 100 %
2. Log in to Artemis with the Student account
     1.  Go the new Programming Exercise
     2. Make a submission that should pass a test (or all) -- You could use the contents from the solution repository.

The updated documentation can be found at https://artemis-platform--8351.org.readthedocs.build/en/8351/admin/setup/programming-exercises.html#caching-maven-dependencies-with-sonatype-nexus

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
Only trivial change in java code.

### CODE REVIEW

**Comment by BaumiCoder** [comment]
Corresponding to the Exam Mode parts of the Pull Request template: Does 
"affecting Programming Exercises" also apply for this Pull Request (i.e. only a template change)?

File: docs/admin/setup/programming-exercises.rst [hunk 1]
```diff
@@ -106,8 +106,147 @@ The ``pipeline.groovy`` file can be customized further by instructors after crea
...
+        --network "artemis-restricted"'
+
+Additionally, on the CI runner host you will have to create the `artemis-restricted` Docker network and some iptables firewall rules to restrict traffic:
+
+.. code-block:: sh
+
+   docker network create --opt com.docker.network.bridge.name=artemis-restr artemis-restricted
+   iptables -I DOCKER-USER -i artemis-restr -j DROP
+   iptables -I DOCKER-USER -i artemis-restr -d $IP_OF_ARTEMIS_EXAMPLE_COM_CACHE -p tcp --dport 8443 -j ACCEPT
+.. _dependecies-docker-volumes:
```
Comment (b-fein) [inline docs/admin/setup/programming-exercises.rst:240]:
```suggestion
   iptables -I DOCKER-USER -i artemis-restr -d $IP_OF_ARTEMIS_EXAMPLE_COM_CACHE -p tcp --dport 8443 -j ACCEPT


.. _dependecies-docker-volumes:
```

File: docs/admin/setup/programming-exercises.rst [hunk 2]
```diff
@@ -106,8 +106,147 @@ The ``pipeline.groovy`` file can be customized further by instructors after crea
...
+
+Caching Maven Dependencies with Sonatype Nexus
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+With Sonatype Nexus you can run a caching server in your local network for Maven dependencies.
+An alternative approach for caching is with docker volumes, see :ref:`dependecies-docker-volumes`.
+
+.. note::
+
+    The following steps assume ``artemis.example.com`` is the host ``8443`` the port and ``10.0.73.42`` for the cache.
```
Comment (b-fein) [inline docs/admin/setup/programming-exercises.rst:119]:
```suggestion
    The following steps assume ``artemis.example.com`` is the host ``10.0.73.42`` and is using port ``8443`` for the cache.
```

**Review by b-fein — APPROVED** [review]
Code/documentation review: Looks good apart from two minor things. :+1:

**Review by b-fein — APPROVED** [review]

**Review by pzdr7 — APPROVED** [review]
demonstrated during testing session; LGTM

**Review by coolchock — APPROVED** [review]
tested during testing session, works as described

**Review by egekurt123 — APPROVED** [review]
Manually tested during testing session on legacy ts2, works as expected

**Review by undernagruzez — APPROVED** [review]
tested during testing session, works as described

**Pull request merged**